tests/lib/smf/src/test_lib_self_transition_smf.c

```diff
@@ -222,6 +220,12 @@ static void parent_ab_run(void *obj)
 	o->transition_bits |= BIT(PARENT_AB_RUN);
 
 	smf_set_state(SMF_CTX(obj), &test_states[STATE_C]);
+	/*
+	 * You should not call smf_set_handled() after smf_set_state().
+	 * There was a bug that did not reset the handled bit if both were called,
+	 * so check it's still fixed:
+	 */
+	smf_set_handled(SMF_CTX(obj));
```

Comment:
You're explicitly violating the API contract by calling `smf_set_handled() after `smf_set_state()`.  

If you swap these calls (`smf_set_handled()` then `smf_set_state()`), does the test still detect the problem?

Reply:
To be honest the API contract should state `smf_set_handled()` is never called in the same code path as `smf_set_state()` or `smf_set_terminate()`, before or after.

Should we error out if people do, or should we accept it as this fix does?

Reply:
The current contract states that it's not required to call `smf_set_handled()` if calling `smf_set_state()`.

You could potentially check if `smf_set_handled()` is called before `smf_set_state()`, but I don't think you could check for the opposite order.

So it's probably best to leave as is, but swap the calling order in the test, as we don't want to have any examples in the tree showing a state doing work after calling `smf_set_state()`.

Reply:
Will do.

Reply:
Confirmed the test fails if `smf_set_handled()` is called before `smf_set_state()` on main.

I'm teaching using SMF, basically telling them "If you handle the event and don't want it propagated to parents, call `smf_set_handled()` except after `smf_set_state()` or `smf_set_terminate()`", and some didn't hear the last bit.

The _correct_ way to fix this would be to put in a return code to state handler functions that returns handled/propagate to parent/ transition occurred/terminate to force people to choose what the state machine does after the handler has run. But doing so would break every existing SMF usage out here, so we're stuck with the `smf_set_handled()` kluge, which either gets forgotten, or put places it's not needed.